hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -1334,6 +1334,29 @@ def draw_bytes(
     ) -> bytes:
         raise NotImplementedError
 
+    def span_start(self, label: int, /) -> None:
```

Comment:
Big fan of the "span" terminology, over "examples". 

In the future we should probably provide a more meaningful representation of the strategy than an integer label, so providers can do particular things for e.g. lists. Pass through the strategy instance wholesale, for examples corresponding to the strategy? I'd guess providers are unlikely to check anything beyond `type(strategy) is ListStrategy`, but could also check attributes like `strategy.min_size` in principle.

Reply:
(maybe we should rename the existing code to match?  big diff but it'd make it easier to work with the internals...)

The `PrimitiveProvider` interface is explicitly unstable for now, so we can change the `label` type when we get around to that ourselves.  Nonetheless IMO passing the strategy instance exposes too much accidental complexity and implementation detail; I do want to at least _try_ evolving towards a clean well-abstracted interface.

Reply:
Pass the strategy type + relevant type-specific kwargs as a dict? 

```python
def span_start(self, label: int, strategy: type[SearchStrategy] | None, strategy_kwargs: dict[str, Any])
```

Agree we can just change it later though, not something we have to decide now. And the label will likely have to stay forever because not all spans are associated with a strategy (eg stateful rules).

-0 on a patch for selfish reasons; I have some local work towards excising all buffer usages, which is very large and will require conflict resolution. But I wouldn't stop such a pull.